gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,285 @@
+    pub fn new(
+        blockstore: Arc<Blockstore>,
+        leader_schedule_cache: Arc<LeaderScheduleCache>,
+    ) -> Self {
+        Self {
+            chunk_map: HashMap::new(),
+            last_root: 0,
+            blockstore,
+            leader_schedule_cache,
+        }
+    }
+
+    fn handle_shred_data(&mut self, data: DuplicateShred) -> Result<(), Error> {
+        if self.should_insert_chunk(&data) {
+            match self.insert_chunk(data) {
+                Err(error) => return Err(error),
+                Ok(Some((slot, proof))) => {
+                    self.verify_and_apply_proof(slot, proof)?;
+                    // We stored the duplicate proof in this slot, no need to accept any future proof.
+                    self.clear_slot(slot);
+                }
+                _ => (),
+            }
+        }
+        Ok(())
+    }
+
+    fn should_insert_chunk(&self, data: &DuplicateShred) -> bool {
+        let slot = data.slot;
+        if slot <= self.blockstore.last_root() || self.blockstore.has_duplicate_shreds_in_slot(slot)
+        {
+            return false;
+        }
+        !matches!(self.chunk_map.get(&slot), Some(None))
+    }
+
+    fn new_proof_chunk_map(num_chunks: u8) -> ProofChunkMap {
+        ProofChunkMap {
+            num_chunks,
+            missing_chunks: num_chunks,
+            chunks: HashMap::new(),
+        }
+    }
+
+    fn clear_slot(&mut self, slot: u64) {
+        self.chunk_map.insert(slot, None);
+    }
+
+    fn insert_chunk(
+        &mut self,
+        data: DuplicateShred,
+    ) -> Result<Option<(Slot, DuplicateSlotProof)>, Error> {
+        if let Some(slot_chunk_map) = self
+            .chunk_map
+            .entry(data.slot)
+            .or_insert_with(|| Some(HashMap::new()))
+        {
+            let proof_key = DuplicateSlotProofKey {
+                from: data.from,
+                wallclock: data.wallclock,
+            };
+            if let Some(proof_chunk_map) = slot_chunk_map
+                .entry(proof_key)
+                .or_insert_with(|| Some(Self::new_proof_chunk_map(data.num_chunks)))
+            {
+                let num_chunks = data.num_chunks;
+                let chunk_index = data.chunk_index;
+                if num_chunks == proof_chunk_map.num_chunks
+                    && chunk_index < num_chunks
+                    && !proof_chunk_map.chunks.contains_key(&chunk_index)
+                {
+                    proof_chunk_map.missing_chunks =
+                        proof_chunk_map.missing_chunks.saturating_sub(1);
+                    proof_chunk_map.chunks.insert(chunk_index, data.chunk);
+                    if proof_chunk_map.missing_chunks == 0 {
+                        let proof_data = (0..num_chunks)
+                            .map(|k| proof_chunk_map.chunks.remove(&k).unwrap())
+                            .concat();
+                        let proof: DuplicateSlotProof = bincode::deserialize(&proof_data)?;
```

Comment:
If somebody deliberately sends us a malicious proof that doesn't deserialize, should we clear the map and retry/ban them for that slot?

Reply:
In that same vein, if somebody sends us a part of a proof, but fails to send the rest of it, should we have a timeout before we clear the chunk so we can ingest somebody else's proof?

Reply:
If the links in the network fail intermittently, it's hard to distinguish between malicious player and network failures. But you raised a good point, we could potentially slash here, but I believe the current Gossip plan is just forwarding duplicate shred proof chunks without double checking the data, so a good player can be tricked into forwarding bad duplicate shred proofs. If we want to totally remove the danger here, we need to guard against the Gossip protocol itself, basically implement slashing in Gossip, it's too late to ban attackers in the layer above Gossip, because we save and forward everything for 30 seconds.

As a stop gap, thinking about the normal cases, we would expect to see 1 or 2 guys sending duplicate shreds, which means we would have ~10 slots of duplicate shred proofs in a rotation of ~2000 in the normal case. I propose only saving ~8 most recent slots of duplicate shred proofs, inside each slot, allow only the most recent proof from each pubkey, also dump things with more than 5 chunks on the floor. So we can limit the size of chunk map to 1500 bytes * 5 * 8 * 2000 = 120MB in the worst case. Does that sound safe enough?

Reply:
Hmm, thought about it more, I guess an attacker can still keep sending us incomplete proofs of fake slots to pollute this map. I'll limit the proof number accepted by each pubkey to 5, then we can limit the map to 1500 bytes * 5 * 5 * 2000 = 75MB.

Reply:
Actually there is no reason for num_chunks to be greater than 3 for now, and by the time the third chunk came we would piece together the full proof. So it's actually 1500 bytes * 2 * 5 * 2000 = 30MB in the worst case. I'm going to insert a per pubkey index to check which slots a pubkey has proof for, this index map would be (256 bytes + 5 * 1) * 2000 = ~600KB in the worst case.

Reply:
done